### PR TITLE
Disable features on tracking opt out

### LINKED ISSUE #6064: Turn off beta features on tracking opt out
As discussed in https://github.com/woocommerce/woocommerce-admin/pull/6054 we need to consider toggling off beta features when opting out of tracking.

We can use the `update_option_woocommerce_allow_tracking` hook to watch and update feature flags.

This might be a nice opportunity to encapsulate features in a way that allows us to more easily detect and toggle feature flags.  E.g., `Feature::get_enabled_toggleable_features()` and `Feature::disable( $feature_slug )`.  Note that the naming between the feature toggle flag (the wp_option) and an "enabled" feature (dev/core/plugin flag) can be confusing.

### PR DESCRIPTION
**Pull request by joshuatf**
Fixes #6064 

* Refactors features and relevant methods to their own file.
* Disables toggleable features on opt out.

### Detailed test instructions:

1. Smoke test features and make sure nothing is broken.
2. Opt in to tracking and the navigation feature.
3. Opt out of tracking.
4. Note the navigation feature is turned off.

### CODE REVIEW

File: src/Features/Navigation/Init.php
Comment (joshuatf) [inline src/Features/Navigation/Init.php:22]: This is a tiny bit magic.  An alternative would be to store the toggleable features in the `Features` class as an associative array.

```php
$toggleable_features = array(
  'navigation' => 'woocommerce_navigation_enabled'
);
```

Though I do like the encapsulation using the current method.

File: src/Loader.php
Comment (joelclimbsthings) [inline src/Loader.php:10]: Looks like this was just obsolete already?

File: src/Features/Features.php
Comment (joelclimbsthings) [inline src/Features/Features.php:110]: So, as it is, any feature class with this `TOGGLE_OPTION_NAME` defined will be disabled with tracking?

File: src/Features/Navigation/Init.php
Comment (joelclimbsthings) [reply src/Features/Navigation/Init.php:22]: I'd lean a bit towards the associative array myself, just for clarity. As it is I wouldn't realize that refactoring/renaming `TOGGLE_OPTION_NAME` within a feature class would disrupt any functionality. 

Alternatively, we could rename that constant to something a little more descriptive, or even add a separate boolean flag that actually determines if the feature is disabled with tracking. ie:
```
const DISABLE_WITH_TRACKING = True;
const TOGGLE_OPTION_NAME = 'woocommerce_navigation_enabled';
```
That way the option name variable would just serve to indicate the option name, and the boolean would fairly clearly indicate the relationship with tracking.

**Review by joelclimbsthings — COMMENTED** [review]
Looks great @joshuatf , definitely love the refactor into `Features.php`! 

Smoke tested the features, and tested the tracking toggle, and no issues found. I added a couple questions on implementation. Interested in your thoughts on those.

File: src/Features/Navigation/Init.php
Comment (joshuatf) [reply src/Features/Navigation/Init.php:22]: Thanks for the feedback here.  I'll update this to use the associative array; I think it does make things more clear.

File: src/Loader.php
Comment (joshuatf) [reply src/Loader.php:10]: It was, I think removed a while ago.

File: src/Features/Features.php
Comment (joshuatf) [reply src/Features/Features.php:110]: That's right.  We'd want to opt out of any toggleable features (presumably all toggleable features are beta features) so this helps with other features going forward.

We may want to rename to `beta_features` as opposed to `toggleable_features` so it's clear that these will be turned off when tracking is opted out.

**Comment by joshuatf** [comment]
@joelclimbsthings Updated this logic in https://github.com/woocommerce/woocommerce-admin/pull/6075/commits/538bbc3ec74349f0958dbe8fd3bfa0ae4e33e4e3

One downside we'll need to be careful of with this method: we need to manually check for the class before adding to the options array as I did in that commit.  It may be easy to slip up and forget that the feature class may or may not be loaded depending on the build.

**Comment by joshuatf** [comment]
Posting here as a reminder to myself:

`is_enabled()` may be better named one of the following to avoid the ambiguity behind being enabled and the feature file being loaded:

* `is_loaded()` - Might not be great since it implies the feature file is already loaded.
* `exists()`
* `has_feature()`
* ...

**Comment by joelclimbsthings** [comment]
Thanks @joshuatf , all looks good and still functions as expected. The only other note I'd add is that it might be a surprise for users that opt out of tracking that the navigation suddenly reverts if they'd forgotten the requirement, or were not the ones to enable navigation in the first place. 

That said, perhaps this is all moot due to the [recent revelation](pbIJXs-wC-p2#comment-1578) that it may be illegal. I'll hold back on approving until we bottom out on that.

**Comment by joshuatf** [comment]
On hold until we determine if we're going to remove the opt-in requirement for beta features.  If we remove that requirement, we'll keep the refactor here less the tracking changes.

**Comment by joelclimbsthings** [comment]
It sounds like [we have a go-ahead](pbIJXs-wC-p2#comment-1647) with this. :ship:

**Review by joelclimbsthings — APPROVED** [review]

**Comment by joshuatf** [comment]
Updated the method names to be more semantic in https://github.com/woocommerce/woocommerce-admin/pull/6075/commits/d291a2bc77de1606f0b58a6ad54f48f9dc27c067 - can you give this a glance @joelclimbsthings or @psealock ?

File: src/Features/Features.php
Comment (psealock) [inline src/Features/Features.php:278]: Should this not use `self:: is_enabled( $key )`?

The previous implementation relied on the logic of the option as well

Comment (joshuatf) [reply src/Features/Features.php]: Good call @psealock.  I think this was not the way the client-side feature flags worked as we still had flag checks on the front-end, but this actually simplifies the logic quite a bit and can't see a reason not to do this right now.

Added this in https://github.com/woocommerce/woocommerce-admin/pull/6075/commits/6cddc360298d5722cf369843e0945e94961f056e and ran through the app to make sure things weren't broken.  Looking good on my end, but a quick sanity check would be nice.

**Comment by joelclimbsthings** [comment]
Just updated as per Paul's requested changes and gave it a quick run through again. Still works. I stand by my approval. :standing_man:

**Pull request merged**
